text/0000-dyno.md

```diff
@@ -0,0 +1,336 @@
+
+Lets examine the changes to `Error` required to make this work:
+
+```rust
+pub mod error {
+    pub trait Error: Debug + Provider {
+        ...
+        fn provide_context<'a>(&'a self, _req: &mut Requisition<'a>) {}
+    }
+
+    impl<T: Error> Provider for T {
+        fn provide<'a>(&'a self, req: &mut Requisition<'a>) {
+            self.provide_context(req);
+        }
+    }
+
+    impl dyn Error + '_ {
+        ...
+        pub fn get_context_ref<T: ?Sized + 'static>(&self) -> Option<&T> {
+            crate::provide_any::request_by_type_tag::<'_, tags::Ref<T>>(self)
+        }
+    }
+}
+```
+
+`get_context_ref` is added as a method on `Error` trait objects (`Error` is also likely to support similar methods for values and possibly other types, but I will elide those details), it simply calls `provide_any::request_by_type_tag` (we'll discuss this function and the `tags::Ref` passed to it below). But where does the context data come from? If a concrete error type supports backtraces, then it must override the `provide_context` method when implementing `Error` (by default, the method does nothing, i.e., no data is provided, so `get_context_ref` will always returns `None`). `provide_context` is used in the blanket implementation of `Provider` for `Error` types, in other words `Provider::provide` is delegated to `Error::provide_context`.
+
+Note that by adding `provide_context` with a default empty implementation and the blanket `impl` of `Provider`, these changes to `Error` are backwards compatible. However, this pattern is only possible because `Provider` and `Error` will be defined in the same crate. For third party users, users will implement `Provider::provide` directly, in the usual way.
+
+In `provide_context`, an error type provides access to data via a `Requisition` object, e.g., `req.provide_ref::<Backtrace>(&self.backtrace)`. The type of the reference passed to `provide_ref` is important here (and we encourage users to use explicit types with turbofish syntax even though it is not necessary, this might even be possible to enforce using a lint). When a user calls `get_context_ref`, the requested type must match the type of an argument to `provide_ref`, e.g., the type of `&self.backtrace` is `&Backtrace`, so a call to `get_context_ref::<Backtrace>()` will return a reference to `self.backtrace`. An implementer can make multiple calls to `provide_ref` to provide multiple data with different types.
+
+Note that `Requisition` has methods for providing values as well as references, and for providing more complex types. These will be covered in the next section.
+
+The important parts of `provide_any` are
+
+```rust
+pub mod provide_any {
+    pub trait Provider {
+        fn provide<'a>(&'a self, req: &mut Requisition<'a>);
```

Comment:
Isn't this unsound now (after `Requisition<'a, '_>` -> `&mut Requisition<'a>` change)?

Since `Requisition` is actually similar to `(TypeId, Option<T>)` for some erased `T`, this theoretically allows writing a `safe_transmute` by 
1. Requesting type `U`
2. In provider replacing request for `U` with a request for `T` via `mem::swap`
3. In provider satisfying the request with `T`-value
4. Getting `U` from the request, but since the request was changed to a request for `T`, this reads `T` as `U`

I couldn't figure out how to get two requests with the same lifetime, but I'm sure there is a way/it's easy to accidentally allow something like this.

It seems like this API should either
1. Use `Pin<&mut Requisition<'a>>`, like the [`object_provider`](https://docs.rs/object-provider/0.2.0/object_provider/index.html) crate
2. Use `Requisition<'a, '_>` and provide a reborrow API to change the `'_` and allow delegating

Reply:
I don't think so because the implementation of `request_by_type_tag` will use the original Requisition (for U) not the new one (for T) so the user will always get back `None`.

Note that the old and new types are equivalent, with the lifetimes fully explicit they are `Requisition<'a, 'b>` and `&'b mut Requisition<'a>`

Reply:
Doesn't `request_by_type_tag` call `provide` with a reference to it's `Requisition`? I imagined it's implemented in a similar facion:
```rust
#[repr(C)]
struct Requisition<'a> {
    id: TypeId,
    ph: PhantomData<&'a ()>,
}

pub fn request_by_type_tag<'a, I: TypeTag<'a>>(provider: &'a dyn Provider) -> Option<I::Type> {
    let mut req = (TypeId::of::<I>(), None::<I::Type>); // this should be a `repr(C)` struct, not a tuple
    let req = transmute::<_, &mut Requisition<'a>>(&mut req);
    provider.provide(req);
    req.1
}
```
In this example `provider` can change the `TypeId` that `Requisition` checks, by replacing it with another `Requisition` with a different `TypeId`, but the slot stays the same.

If I understand correctly, `Requisition<'a, 'b>` was wrapping `*mut` inside and so replacing it wouldn't bring inconsistency between `TypeId` and the slot.

Reply:
Ah, I see what you mean, sorry.

Implementation is here: https://github.com/rust-lang/rust/pull/91970/files#diff-0752889661748b8a15a597d7156127b6fb90fdeda0627be50e07f3f785bd0f4dR798-R805

Requisition never stores the type id, it is synthesised as needed.

Its not actually possible to create a new Requisition since TagValue is private to the any module. In addition, Requisition is unsized so I think that even if you could create one, you can't assign it into the mutable reference.

Reply:
> Its not actually possible to create a new Requisition since TagValue is private to the any module

If `Requisition` was `Sized`, the only thing you'd need would be `&mut Requisition`, since you could use `mem::swap`.

> Implementation is here: https://github.com/rust-lang/rust/pull/91970/files#diff-`0752889661748b8a15a597d7156127b6fb90fdeda0627be50e07f3f785bd0f4dR798-R805

I see. Interesting, thanks!

Reply:
> If Requisition was Sized, the only thing you'd need would be &mut Requisition, since you could use mem::swap

I don't understand, how would you create the new Requisition to swap in? Requisition has a TagValue field and you need to name that to create it and thus create the Requisition.

Reply:
You could do another `request_by_type_tag`, but I suppose there's no way to get the lifetimes to match up.